File: postlio_backend/app/api/v1/brands.py
# postlio_backend/app/api/v1/brands.py
import json
import re
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, status, UploadFile, File
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db, get_current_user
from app.api.exceptions import NotFoundError
from app.models.user import User
from app.models.brand import Brand
from app.repositories import brand_repo
from app.schemas.brand import (
    BrandCreate,
    BrandUpdate,
    BrandResponse,
    BrandsListResponse,
    AnalyzeVoiceRequest,
    VoiceAnalysisResult,
    BrandAnalytics,
    VoiceDNABase,
)
from app.services.ai.text.manager import TextAIManager
from app.services.storage import r2_storage

MAX_LOGO_SIZE_BYTES = 5 * 1024 * 1024

router = APIRouter(prefix="/brands", tags=["Brand Voice"])



@router.get("/", response_model=BrandsListResponse)
async def get_brands(
        db: AsyncSession = Depends(get_db),
        current_user: User = Depends(get_current_user),
        skip: int = 0,
        limit: int = 50,
        is_active: Optional[bool] = None,
):
    brands = await brand_repo.list_brands(db, current_user.id, is_active, skip, limit)
    total = await brand_repo.count_brands(db, current_user.id, is_active)
    return BrandsListResponse(brands=[BrandResponse.model_validate(b) for b in brands], total=total)


@router.get("/{brand_id}", response_model=BrandResponse)
async def get_brand(
        brand_id: int,
        db: AsyncSession = Depends(get_db),
        current_user: User = Depends(get_current_user),
):
    brand = await brand_repo.get_by_id(db, current_user.id, brand_id)
    if not brand:
        raise NotFoundError("Brand")
    return BrandResponse.model_validate(brand)


@router.post("/", response_model=BrandResponse, status_code=status.HTTP_201_CREATED)
async def create_brand(
        data: BrandCreate,
        db: AsyncSession = Depends(get_db),
        current_user: User = Depends(get_current_user),
):
    voice_dna_dict = data.voice_dna.model_dump() if data.voice_dna else VoiceDNABase().model_dump()
    is_first = await brand_repo.count_brands(db, current_user.id) == 0

    brand = Brand(
        user_id=current_user.id,
        name=data.name,
        description=data.description,
        logo_url=data.logo_url,
        primary_color=data.primary_color,
        secondary_color=data.secondary_color,
        industry=data.industry,
        target_audience=data.target_audience,
        voice_dna=voice_dna_dict,
        is_default=is_first,
    )
    brand = await brand_repo.create(db, brand)
    return BrandResponse.model_validate(brand)


@router.patch("/{brand_id}", response_model=BrandResponse)
async def update_brand(
        brand_id: int,
        data: BrandUpdate,
        db: AsyncSession = Depends(get_db),
        current_user: User = Depends(get_current_user),
):
    brand = await brand_repo.get_by_id(db, current_user.id, brand_id)
    if not brand:
        raise NotFoundError("Brand")

    for field, value in data.model_dump(exclude_unset=True, exclude={"voice_dna"}).items():
        setattr(brand, field, value)

    if data.voice_dna:
        current_voice_dna = brand.voice_dna or {}
        current_voice_dna.update(data.voice_dna.model_dump(exclude_unset=True))
        brand.voice_dna = current_voice_dna

    brand = await brand_repo.save(db, brand)
    return BrandResponse.model_validate(brand)


@router.delete("/{brand_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_brand(
        brand_id: int,
        db: AsyncSession = Depends(get_db),
        current_user: User = Depends(get_current_user),
):
    brand = await brand_repo.get_by_id(db, current_user.id, brand_id)
    if not brand:
        raise NotFoundError("Brand")

    was_default = brand.is_default
    await brand_repo.delete(db, brand)

    if was_default:
        new_default = await brand_repo.get_latest(db, current_user.id)
        if new_default:
            new_default.is_default = True
            await brand_repo.save(db, new_default)


@router.post("/{brand_id}/set-default", response_model=BrandResponse)
async def set_default_brand(
        brand_id: int,
        db: AsyncSession = Depends(get_db),
        current_user: User = Depends(get_current_user),
):
    brand = await brand_repo.get_by_id(db, current_user.id, brand_id)
    if not brand:
        raise NotFoundError("Brand")

    await brand_repo.clear_default(db, current_user.id)
    brand.is_default = True
    brand = await brand_repo.save(db, brand)
    return BrandResponse.model_validate(brand)


@router.post("/{brand_id}/logo")
async def upload_brand_logo(
        brand_id: int,
        file: UploadFile = File(...),
        db: AsyncSession = Depends(get_db),
        current_user: User = Depends(get_current_user),
):
    brand = await brand_repo.get_by_id(db, current_user.id, brand_id)
    if not brand:
        raise NotFoundError("Brand")

    if not file.content_type or not file.content_type.startswith("image/"):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Plik musi być obrazem")

    image_bytes = await file.read()
    if len(image_bytes) > MAX_LOGO_SIZE_BYTES:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Plik jest za duży (max {MAX_LOGO_SIZE_BYTES // (1024 * 1024)}MB)",
        )

    logo_url = await r2_storage.upload_image(image_bytes, folder="brand-logos")
    if not logo_url:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Nie udało się zapisać logo. Spróbuj ponownie później.",
        )

    brand.logo_url = logo_url
    await brand_repo.save(db, brand)
    return {"logo_url": logo_url}


@router.get("/{brand_id}/analytics", response_model=BrandAnalytics)
async def get_brand_analytics(
        brand_id: int,
        db: AsyncSession = Depends(get_db),
        current_user: User = Depends(get_current_user),
):
    brand = await brand_repo.get_by_id(db, current_user.id, brand_id)
    if not brand:
        raise NotFoundError("Brand")

    return BrandAnalytics(
        total_posts=brand.posts_count or 0,
        posts_this_month=0,
        engagement_rate=0.0,
        best_performing_platform=None,
        suggested_posting_times=["10:00", "14:00", "18:00"],
    )


@router.post("/analyze-voice", response_model=VoiceAnalysisResult)
async def analyze_brand_voice(
        data: AnalyzeVoiceRequest,
        current_user: User = Depends(get_current_user),
):
    if not data.sample_content:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Wymagany co najmniej 1 przykładowy tekst")

    combined_content = "\n---\n".join(data.sample_content)
    analysis_prompt = f"""Przeanalizuj poniższe przykładowe treści marki i określ jej głos (Voice DNA).

TREŚCI DO ANALIZY:
{combined_content}

Odpowiedz w formacie JSON:
{{
    "tone_formality": <0-100>,
    "tone_energy": <0-100>,
    "tone_humor": <0-100>,
    "tone_emotion": <0-100>,
    "personality_traits": [<lista 3-5 cech>],
    "communication_style": <jeden z: informative, inspirational, educational, entertaining, conversational, storytelling>,
    "detected_keywords": [<5-10 słów kluczowych>],
    "detected_hashtags": [<hashtagi jeśli wykryte>],
    "emoji_usage": <none, minimal, moderate, frequent>,
    "vocabulary_level": <prosty, średni, zaawansowany, ekspercki>,
    "common_phrases": [<3-5 charakterystycznych fraz>]
}}

Odpowiedz TYLKO JSON, bez dodatkowego tekstu."""

    try:
        ai_manager = TextAIManager()
        ai_response = await ai_manager.generate(
            prompt=analysis_prompt,
            system_prompt="Jesteś ekspertem od analizy marki i copywritingu.",
        )

        json_match = re.search(r"\{[\s\S]*\}", ai_response)
        if json_match:
            analysis_data = json.loads(json_match.group())
        else:
            raise ValueError("Nie znaleziono JSON w odpowiedzi AI")

        voice_dna = VoiceDNABase(
            tone_formality=analysis_data.get("tone_formality", 50),
            tone_energy=analysis_data.get("tone_energy", 50),
            tone_humor=analysis_data.get("tone_humor", 30),
            tone_emotion=analysis_data.get("tone_emotion", 50),
            personality_traits=analysis_data.get("personality_traits", ["professional", "friendly"]),
            communication_style=analysis_data.get("communication_style", "informative"),
            keywords=analysis_data.get("detected_keywords", []),
            hashtags=analysis_data.get("detected_hashtags", []),
            emoji_usage=analysis_data.get("emoji_usage", "moderate"),
        )
        analysis = {
            "tone_breakdown": {
                "formality": analysis_data.get("tone_formality", 50),
                "energy": analysis_data.get("tone_energy", 50),
                "humor": analysis_data.get("tone_humor", 30),
                "emotion": analysis_data.get("tone_emotion", 50),
            },
            "common_phrases": analysis_data.get("common_phrases", []),
            "vocabulary_level": analysis_data.get("vocabulary_level", "średni"),
            "emoji_usage_detected": analysis_data.get("emoji_usage", "moderate"),
            "hashtag_style": "branded" if analysis_data.get("detected_hashtags") else "minimal",
        }
        return VoiceAnalysisResult(voice_dna=voice_dna, analysis=analysis)

    except json.JSONDecodeError as e:
        raise HTTPException(status_code=500, detail=f"Błąd parsowania odpowiedzi AI: {e}")
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Błąd analizy: {e}")

File: postlio_frontend/src/lib/api/brands.ts
// src/lib/api/brands.ts
import { apiClient } from './client';
import type {
    Brand,
    BrandVoiceDNA,
    ApiBrand,
    ApiBrandsListResponse,
    ApiAnalyzeVoiceResponse,
} from '@/types/brand';
import {
    transformApiBrandToFrontend,
    transformApiVoiceDNAToFrontend,
    transformVoiceDNAToApi,
} from '@/types/brand';


// ============================================================
// REQUEST TYPES
// ============================================================

export interface CreateBrandRequest {
    name: string;
    description?: string;
    logoUrl?: string;
    primaryColor?: string;
    secondaryColor?: string;
    industry?: string;
    targetAudience?: string;
    voiceDNA?: Partial<BrandVoiceDNA>;
}

export interface UpdateBrandRequest {
    name?: string;
    description?: string;
    logoUrl?: string;
    primaryColor?: string;
    secondaryColor?: string;
    industry?: string;
    targetAudience?: string;
    voiceDNA?: Partial<BrandVoiceDNA>;
    isActive?: boolean;
}

export interface AnalyzeBrandVoiceRequest {
    sample_content: string[];
    brand_id?: number;
}

// ============================================================
// RESPONSE TYPES (frontend)
// ============================================================

export interface BrandsListResponse {
    brands: Brand[];
    total: number;
}

export interface AnalyzeBrandVoiceResponse {
    voiceDNA: BrandVoiceDNA;
    analysis: {
        toneBreakdown: Record<string, number>;
        commonPhrases: string[];
        vocabularyLevel: string;
        emojiUsageDetected: string;
        hashtagStyle: string;
    };
}

export interface BrandAnalytics {
    totalPosts: number;
    postsThisMonth: number;
    engagementRate: number;
    bestPerformingPlatform: string | null;
    suggestedPostingTimes: string[];
}

// ============================================================
// API FUNCTIONS
// ============================================================

/**
 * Pobierz listę brandów
 */
export async function getBrands(): Promise<BrandsListResponse> {
    const response = await apiClient.get<ApiBrandsListResponse>('/brands');
    return {
        brands: response.brands.map(transformApiBrandToFrontend),
        total: response.total,
    };
}

/**
 * Pobierz pojedynczy brand
 */
export async function getBrand(id: string): Promise<Brand> {
    const response = await apiClient.get<ApiBrand>(`/brands/${id}`);
    return transformApiBrandToFrontend(response);
}

/**
 * Utwórz nowy brand
 */
export async function createBrand(data: CreateBrandRequest): Promise<Brand> {
    const apiData = {
        name: data.name,
        description: data.description,
        logo_url: data.logoUrl,
        primary_color: data.primaryColor || '#8B5CF6',
        secondary_color: data.secondaryColor,
        industry: data.industry,
        target_audience: data.targetAudience,
        voice_dna: data.voiceDNA ? transformVoiceDNAToApi(data.voiceDNA) : undefined,
    };

    const response = await apiClient.post<ApiBrand>('/brands', apiData);
    return transformApiBrandToFrontend(response);
}

/**
 * Aktualizuj brand
 */
export async function updateBrand(
    id: string,
    data: UpdateBrandRequest
): Promise<Brand> {
    const apiData: Record<string, unknown> = {};

    if (data.name !== undefined) apiData.name = data.name;
    if (data.description !== undefined) apiData.description = data.description;
    if (data.logoUrl !== undefined) apiData.logo_url = data.logoUrl;
    if (data.primaryColor !== undefined) apiData.primary_color = data.primaryColor;
    if (data.secondaryColor !== undefined) apiData.secondary_color = data.secondaryColor;
    if (data.industry !== undefined) apiData.industry = data.industry;
    if (data.targetAudience !== undefined) apiData.target_audience = data.targetAudience;
    if (data.isActive !== undefined) apiData.is_active = data.isActive;
    if (data.voiceDNA !== undefined) apiData.voice_dna = transformVoiceDNAToApi(data.voiceDNA);

    const response = await apiClient.patch<ApiBrand>(`/brands/${id}`, apiData);
    return transformApiBrandToFrontend(response);
}

/**
 * Usuń brand
 */
export async function deleteBrand(id: string): Promise<void> {
    return apiClient.delete(`/brands/${id}`);
}

/**
 * Upload logo brandu
 */
export async function uploadBrandLogo(
    brandId: string,
    file: File,
    onProgress?: (percent: number) => void
): Promise<{ logoUrl: string }> {
    const response = await apiClient.upload<{ logo_url: string }>(
        `/brands/${brandId}/logo`,
        file,
        onProgress
    );
    return { logoUrl: response.logo_url };
}

/**
 * Analizuj Brand Voice na podstawie przykładowych treści
 */
export async function analyzeBrandVoice(
    data: AnalyzeBrandVoiceRequest
): Promise<AnalyzeBrandVoiceResponse> {
    const response = await apiClient.post<ApiAnalyzeVoiceResponse>(
        '/brands/analyze-voice',
        data,
        { timeout: 45000 }
    );

    return {
        voiceDNA: transformApiVoiceDNAToFrontend(response.voice_dna),
        analysis: {
            toneBreakdown: response.analysis.tone_breakdown,
            commonPhrases: response.analysis.common_phrases,
            vocabularyLevel: response.analysis.vocabulary_level,
            emojiUsageDetected: response.analysis.emoji_usage_detected,
            hashtagStyle: response.analysis.hashtag_style,
        },
    };
}

/**
 * Pobierz analityki brandu
 */
export async function getBrandAnalytics(id: string): Promise<BrandAnalytics> {
    const response = await apiClient.get<{
        total_posts: number;
        posts_this_month: number;
        engagement_rate: number;
        best_performing_platform: string | null;
        suggested_posting_times: string[];
    }>(`/brands/${id}/analytics`);

    return {
        totalPosts: response.total_posts,
        postsThisMonth: response.posts_this_month,
        engagementRate: response.engagement_rate,
        bestPerformingPlatform: response.best_performing_platform,
        suggestedPostingTimes: response.suggested_posting_times,
    };
}

/**
 * Ustaw brand jako domyślny
 */
export async function setDefaultBrand(id: string): Promise<Brand> {
    const response = await apiClient.post<ApiBrand>(`/brands/${id}/set-default`);
    return transformApiBrandToFrontend(response);
}

// ============================================================
// EXPORT ZBIORCZY
// ============================================================

export const brandsApi = {
    getBrands,
    getBrand,
    createBrand,
    updateBrand,
    deleteBrand,
    uploadBrandLogo,
    analyzeBrandVoice,
    getBrandAnalytics,
    setDefaultBrand,
};

export default brandsApi;